File: core/src/main/java/io/github/msameer0/rhythmicrush/game/level/ProgressManager.java
package io.github.msameer0.rhythmicrush.game.level;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.files.FileHandle;
import com.badlogic.gdx.utils.Json;
import com.badlogic.gdx.utils.JsonValue;
import com.badlogic.gdx.utils.JsonWriter;
import com.badlogic.gdx.utils.ObjectMap;


/**
 * Manages the persistence of player progress for different levels.
 * Handles the loading and saving of {@link LevelProgress} data to a local JSON file.
 */
public class ProgressManager {

    private static final String SAVE_PATH = "saves/progress.json";

    private final ObjectMap<String, LevelProgress> map = new ObjectMap<>();
    private final Json json;

    public ProgressManager() {
        json = new Json();
        json.setOutputType(JsonWriter.OutputType.json);
        json.setUsePrototypes(false);
        load();
    }


    public LevelProgress getOrCreate(String levelKey) {
        if (!map.containsKey(levelKey)) {
            map.put(levelKey, new LevelProgress());
        }
        return map.get(levelKey);
    }

    /**
     * Persists the current player progress to a local JSON file.
     * <p>
     * This method serializes the internal map of {@link LevelProgress} objects into a JSON format
     * and writes it to the local storage path defined by {@code SAVE_PATH}. If the parent
     * directories do not exist, they are created automatically. In the event of an IO failure
     * or serialization error, the exception is caught and logged via {@link Gdx#app}.
     * </p>
     */
    public void save() {
        try {
            FileHandle file = Gdx.files.local(SAVE_PATH);
            file.parent().mkdirs();

            StringBuilder sb = new StringBuilder("{\n");
            boolean first = true;
            for (ObjectMap.Entry<String, LevelProgress> entry : map) {
                if (!first) sb.append(",\n");
                first = false;
                sb.append("  \"").append(entry.key).append("\": ");
                sb.append(json.toJson(entry.value));
            }
            sb.append("\n}");
            file.writeString(sb.toString(), false);
        } catch (Exception e) {
            Gdx.app.error("ProgressManager", "Failed to save progress: " + e.getMessage());
        }
    }


    /**
     * Loads the player progress from the local storage file.
     * <p>
     * This method attempts to read the JSON file at {@code SAVE_PATH}. If the file exists,
     * it parses the content and populates the internal progress map. If the file does not
     * exist, the method returns silently, leaving the map empty. Any exceptions encountered
     * during the reading or parsing process are caught and logged.
     * </p>
     */
    private void load() {
        try {
            FileHandle file = Gdx.files.local(SAVE_PATH);
            if (!file.exists()) return;

            JsonValue root = new com.badlogic.gdx.utils.JsonReader().parse(file);
            for (JsonValue entry = root.child; entry != null; entry = entry.next) {
                LevelProgress p = json.readValue(LevelProgress.class, entry);
                if (p != null) map.put(entry.name, p);
            }
        } catch (Exception e) {
            Gdx.app.error("ProgressManager", "Failed to load progress: " + e.getMessage());
        }
    }
}

File: core/src/main/kotlin/io/github/msameer0/rhythmicrush/game/level/ProgressManager.kt
package io.github.msameer0.rhythmicrush.game.level

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.utils.Json
import com.badlogic.gdx.utils.JsonWriter
import com.badlogic.gdx.utils.ObjectMap
import com.badlogic.gdx.utils.UBJsonReader
import com.badlogic.gdx.utils.UBJsonWriter
import io.github.msameer0.rhythmicrush.account.CloudProgress
import io.github.msameer0.rhythmicrush.account.LevelCloudProgress
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * Manages the persistence of level completion progress and attempt counts.
 */
class ProgressManager {
    companion object {
        const val SAVE_PATH: String = "saves/progress.ubj"
        const val COINS_KEY: String = "coins"
        const val POINTS_KEY: String = "points"
        const val SCHEMA_KEY: String = "schemaVersion"
        const val CURRENT_SCHEMA: Int = 2
        private const val BACKUP_DIRECTORY = "saves/backups"
        private const val MAX_BACKUPS = 5
    }

    val map = ObjectMap<String, LevelProgress>()
    private val json = Json()
    var coins: Int = 0
    var points: Int = 0

    init {
        json.setOutputType(JsonWriter.OutputType.json)
        json.setUsePrototypes(false)
        load()
    }

    fun getOrCreate(levelKey: String): LevelProgress {
        val existingProgress = map.get(levelKey)
        if (existingProgress != null) return existingProgress

        return LevelProgress().also { map.put(levelKey, it) }
    }

    fun save(createBackup: Boolean = false) {
        Gdx.app.log("ProgressManager", "Saving progress...")
        try {
            val file = Gdx.files.local(SAVE_PATH)
            file.parent().mkdirs()
            val temporary = Gdx.files.local("$SAVE_PATH.tmp")
            writeSnapshot(temporary)
            UBJsonReader().parse(temporary)
            if (createBackup && file.exists()) createBackup(file)
            val previous = Gdx.files.local("$SAVE_PATH.previous")
            if (previous.exists()) previous.delete()
            if (file.exists()) file.moveTo(previous)
            try {
                temporary.moveTo(file)
                if (previous.exists()) previous.delete()
            } catch (exception: Exception) {
                if (!file.exists() && previous.exists()) previous.moveTo(file)
                throw exception
            }
            Gdx.app.log("ProgressManager", "Progress saved successfully.")
        } catch (exception: Exception) {
            Gdx.app.error(
                "ProgressManager",
                "Failed to save progress: ${exception.message}"
            )
        }
    }

    fun exportForSync(): List<LevelCloudProgress> {
        val result = mutableListOf<LevelCloudProgress>()
        for (entry in map) {
            result.add(
                LevelCloudProgress(
                    levelId = entry.key,
                    bestPercent = entry.value.bestPercent,
                    totalAttempts = entry.value.localDeviceAttempts,
                    completed = entry.value.bestPercent >= 100
                )
            )
        }
        return result
    }

    fun mergeCloud(progress: CloudProgress) {
        for (cloud in progress.levels) {
            val local = getOrCreate(cloud.levelId)
            local.bestPercent = maxOf(local.bestPercent, cloud.bestPercent)
            local.totalAttempts = maxOf(local.totalAttempts, cloud.totalAttempts)
            if (cloud.completed) local.completionRewardGranted = true
        }
        coins = progress.coins
        points = progress.points
        save(createBackup = true)
    }

    private fun load() {
        Gdx.app.log("ProgressManager", "Loading progress...")
        try {
            val file = Gdx.files.local(SAVE_PATH)
            if (!file.exists()) {
                Gdx.app.log("ProgressManager", "No progress file found.")
                return
            }

            val root = UBJsonReader().parse(file)
            val savedSchema = root.getInt(SCHEMA_KEY, 1)
            var entry = root.child
            while (entry != null) {
                when (entry.name) {
                    SCHEMA_KEY -> Unit
                    COINS_KEY -> coins = entry.asInt()
                    POINTS_KEY -> points = entry.asInt()
                    else -> {
                        val levelProgress =
                            json.readValue<LevelProgress?>(
                                LevelProgress::class.java,
                                entry
                            )
                        if (levelProgress != null) {
                            if (savedSchema < 2) {
                                levelProgress.localDeviceAttempts = levelProgress.totalAttempts
                                if (levelProgress.bestPercent >= 100) {
                                    levelProgress.completionRewardGranted = true
                                }
                            }
                            map.put(entry.name, levelProgress)
                        }
                    }
                }
                entry = entry.next
            }
            Gdx.app.log("ProgressManager", "Progress loaded: " + map.size + " entries.")
        } catch (exception: Exception) {
            Gdx.app.error(
                "ProgressManager",
                "Failed to load progress: ${exception.message}"
            )
        }
    }

    private fun writeSnapshot(file: com.badlogic.gdx.files.FileHandle) {
        val writer = UBJsonWriter(file.write(false))
        try {
            writer.`object`()
            writer.set(SCHEMA_KEY, CURRENT_SCHEMA)
            writer.set(COINS_KEY, coins)
            writer.set(POINTS_KEY, points)
            for (entry in map) {
                writer.name(entry.key)
                writer.value(com.badlogic.gdx.utils.JsonReader().parse(json.toJson(entry.value)))
            }
            writer.pop()
        } finally {
            writer.close()
        }
    }

    private fun createBackup(source: com.badlogic.gdx.files.FileHandle) {
        val directory = Gdx.files.local(BACKUP_DIRECTORY)
        directory.mkdirs()
        val timestamp = SimpleDateFormat("yyyyMMdd-HHmmss-SSS", Locale.US).format(Date())
        source.copyTo(directory.child("progress-$timestamp.ubj"))
        val backups = directory.list(".ubj").sortedByDescending { it.lastModified() }
        for (index in MAX_BACKUPS until backups.size) backups[index].delete()
    }
}
